libopenage/unit/producer.cpp

```diff
@@ -614,17 +643,33 @@ TerrainObject *BuildingProducer::place(Unit *u, std::shared_ptr<Terrain> terrain
 	 */
 	TerrainObject *obj_ptr = u->location.get();
 	std::weak_ptr<Terrain> terrain_ptr = terrain;
-	u->location->passable = [obj_ptr, terrain_ptr](const coord::phys3 &pos) -> bool {
+
+    // find set of allowed terrains
+    std::unordered_set<terrain_t> terrains = allowed_terrains(this->unit_data.terrain_restriction);
+
+    u->location->passable = [obj_ptr, terrain_ptr, terrains](const coord::phys3 &pos) -> bool {
```

Comment:
It's a good day when indentation is my worst problem :) I'm going to try finishing another feature (building docks) and then fix all of my changes per the style guide for merging